libevm/lib/handles.go

```diff
@@ -0,0 +1,65 @@
+package lib
+
+import (
+	"errors"
+	"fmt"
+	"math"
+	"sync"
+)
+
+var ErrInvalidHandle = errors.New("invalid handle")
+
+type Handles[T comparable] struct {
+	used    map[int]T
+	current int
+	mutex   *sync.RWMutex
+}
+
+func NewHandles[T comparable]() *Handles[T] {
+	return &Handles[T]{
+		used:    make(map[int]T),
+		current: 0,
+		mutex:   &sync.RWMutex{},
+	}
+}
+
+func (h *Handles[T]) Add(obj T) int {
+	h.mutex.Lock()
+	defer h.mutex.Unlock()
+	// pedantic safety check for overflow: this is highly unlikely because there are 2**31-1 available handles,
+	// just the memory consumption alone will likely cause problems before this happens
+	if len(h.used) == math.MaxInt32 {
```

Comment:
i'm not sure about this check.
From the code below:
```
if h.current == math.MaxInt32 {
		h.current = 0
}
h.current++
```
I see, that it is possible to have `current == MaxInt32` -> that will lead to `len(h.used)` to be exactly `math.MaxInt32`

I think, we need to change the code below to `if h.current == math.MaxInt32 - 1`
or something like this

Reply:
> I see, that it is possible to have current == MaxInt32 -> that will lead to len(h.used) to be exactly math.MaxInt32

Correct, in that situation the next item doesnt fit anymore, which is why the check is for `len(h.used) == math.MaxInt32`. I think it is ok like this. The system will most likely be out of memory at this point anyway.

Reply:
hes sense!
thanks